shaders/src/light_reflections.fs

```diff
@@ -0,0 +1,276 @@
+
+mat4x4 scaleMatrix(float x, float y) {
+    mat4x4 m = mat4(1.0f);
+    m[0].x = x;
+    m[1].y = y;
+    m[2].z = 1.0f;
+    m[3].w = 1.0f;
+    return m;
+}
+
+mat4x4 translateMatrix(float x, float y) {
+    mat4x4 m = mat4(1.0f);
+    m[3].xy = vec2(x, y);
+    return m;
+}
+
+// random number between 0 and 1, using interleaved gradient noise
+float ssr_random(const highp vec2 w) {
+    const vec3 m = vec3(0.06711056, 0.00583715, 52.9829189);
+    return fract(m.z * fract(dot(w, m.xy)));
+}
+
+/**
+ * Evaluates screen-space reflections, storing the color in Fr if there's a hit.
+ * r is the desired reflected vector.
+ * Returns true if there's a hit, false otherwise.
+ */
+bool evaluateScreenSpaceReflections(highp vec3 r, inout vec3 Fr) {
+    highp vec3 wsRayDirection = r;
+    highp vec3 wsRayStart = shading_position + frameUniforms.ssrBias * wsRayDirection;
+
+    // ray start/end in view space
+    highp vec4 vsRayStart = getViewFromWorldMatrix() * vec4(wsRayStart, 1.0);
+    highp vec4 vsRayDirection = getViewFromWorldMatrix() * vec4(wsRayDirection, 0.0);
+
+    highp vec3 vsOrigin = vsRayStart.xyz;
+    highp vec3 vsDirection = vsRayDirection.xyz;
+    highp float vsZThickness = frameUniforms.ssrThickness;
+    // TODO: use the actual near plane.
+    highp float nearPlaneZ = -0.1f;
+    highp float stride = 1.0;
+    // TODO: jitterFraction should be a between 0 and 1, but anything < 1 gives banding artifacts.
+    highp float jitterFraction = 1.0f;
+    // TODO: set this to the larger of the viewport dimensions.
+    highp float maxSteps = 1000.0;
+    highp float maxRayTraceDistance = frameUniforms.ssrDistance;
+
+    vec2 res = vec2(textureSize(light_structure, 0).xy);
+    highp mat4x4 projectToPixelMatrix =
+        scaleMatrix(res.x, res.y) *
+        translateMatrix(0.5f, 0.5f) *
+#if defined(TARGET_METAL_ENVIRONMENT) || defined(TARGET_VULKAN_ENVIRONMENT)
+        scaleMatrix(0.5f, -0.5f) *  // Metal and Vulkan's texture space is y-inverted
+#else
+        scaleMatrix(0.5f, 0.5f) *
+#endif
+        getClipFromViewMatrix();
+
+    // Outputs from the traceScreenSpaceRay function.
+    vec2 hitPixel;  // not currently used
+    vec3 vsHitPoint;
+
+    if (traceScreenSpaceRay(vsOrigin, vsDirection, projectToPixelMatrix, light_structure,
+            vsZThickness, nearPlaneZ, stride, jitterFraction, maxSteps,
+            maxRayTraceDistance, hitPixel, vsHitPoint)) {
+        vec2 ssrRes = vec2(textureSize(light_ssr, 0).xy);
+        vec4 reprojected = scaleMatrix(ssrRes.x, ssrRes.y) *
+                translateMatrix(0.5f, 0.5f) *
+#if defined(TARGET_METAL_ENVIRONMENT) || defined(TARGET_VULKAN_ENVIRONMENT)
+                scaleMatrix(0.5f, -0.5f) *
+#else
+                scaleMatrix(0.5f, 0.5f) *
+#endif
+                frameUniforms.ssrReprojection *
+                vec4(vsHitPoint, 1.0f);
+        reprojected *= (1.0 / reprojected.w);
+        Fr = texelFetch(light_ssr, int2(reprojected.x, reprojected.y), 0).rgb;
```

Comment:
We should fade the contribution of the reflection near the edges, to avoid a sharp cutoff (See screenshot).
![Screen Shot 2021-12-12 at 11 15 52 PM](https://user-images.githubusercontent.com/1240896/145768515-67d93841-ee15-44b1-875f-9ddcd06fa491.png)

We already do something like that for the contact shadows, this is the code we use there:
```
    vec2 fade = max(12.0 * abs(ray.xy - 0.5) - 5.0, 0.0);
    occlusion *= saturate(1.0 - dot(fade, fade));
```
This can be adapted here, using a mix().

We should also probably fade based on the max ray-length vs hitpoint.

Reply:
Why use `texelFetch` here? This prevents using bilinear filtering and causes issues at the edges. Use `textureLod` instead, the only modification to the code is to not scale by `ssrRes` when calculating `reprojected` above.
This will also help with the fading mentioned above.